src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -50,13 +105,121 @@ rawPageNumBlobs :: RawPage -> Word16
+rawPageValueOffsets :: RawPage -> P.Vector ValueOffset
+rawPageValueOffsets page@(RawPage off ba) =
+    P.Vector (off + fromIntegral (div2 dirOffset) + fromIntegral dirNumKeys) (fromIntegral dirNumKeys + 1) ba
+  where
+    Dir {..} = rawPageDirectory page
+
+-- | single key page case
+rawPageValueOffsets1 :: RawPage -> (Word16, Word32)
+rawPageValueOffsets1 page@(RawPage off ba) =
+    ( indexByteArray ba (off + fromIntegral (div2 dirOffset) + 1)
+    , indexByteArray ba (div2 (off + fromIntegral (div2 dirOffset)) + 1)
+    )
+  where
+    Dir {..} = rawPageDirectory page
+
+rawPageHasBlobRefAt :: RawPage -> Int -> Word64
+rawPageHasBlobRefAt _page@(RawPage off ba) i = do
+    let j = unsafeShiftR i 6 -- `div` 64
+    let k = i .&. 63         -- `mod` 64
+    let word = indexByteArray ba (div4 off + 1 + j)
+    unsafeShiftR word k .&. 1
+
+rawPageOpAt :: RawPage -> Int -> Word64
+rawPageOpAt page@(RawPage off ba) i = do
+    let j = unsafeShiftR i 5 -- `div` 32
+    let k = i .&. 31         -- `mod` 32
+    let word = indexByteArray ba (div4 off + 1 + roundUpTo64 (fromIntegral dirNumKeys) + j)
+    unsafeShiftR word (mul2 k) .&. 3
+  where
+    Dir {..} = rawPageDirectory page
+
+roundUpTo64 :: Int -> Int
+roundUpTo64 i =
+    if k == 0 then j else j + 1
+  where
+    j = unsafeShiftR i 6 -- `div` 64
+    k = i .&. 63         -- `mod` 64
+
+rawPageKeys :: RawPage -> V.Vector (P.Vector Word8)
+rawPageKeys page@(RawPage off ba) = do
+    let offs = rawPageKeyOffsets page
+    V.fromList
+        [ P.Vector (mul2 off + start) (end - start) ba
+        | i <- [ 0 .. fromIntegral dirNumKeys -  1 ] :: [Int]
+        , let start = fromIntegral (P.unsafeIndex offs i) :: Int
+        , let end   = fromIntegral (P.unsafeIndex offs (i + 1)) :: Int
+        ]
+  where
+    Dir {..} = rawPageDirectory page
+
+rawPageKeyAt :: RawPage -> Int -> P.Vector Word8
+rawPageKeyAt page@(RawPage off ba) i = do
+    P.Vector (mul2 off + start) (end - start) ba
+  where
+    offs  = rawPageKeyOffsets page
+    start = fromIntegral (P.unsafeIndex offs i) :: Int
+    end   = fromIntegral (P.unsafeIndex offs (i + 1)) :: Int
+
+-- | Non-single page case
+rawPageValues :: RawPage -> V.Vector (P.Vector Word8)
+rawPageValues page@(RawPage off ba) = do
+    let offs = rawPageValueOffsets page
+    V.fromList
+        [ P.Vector (mul2 off + start) (end - start) ba
+        | i <- [ 0 .. fromIntegral dirNumKeys -  1 ] :: [Int]
+        , let start = fromIntegral (P.unsafeIndex offs i) :: Int
+        , let end   = fromIntegral (P.unsafeIndex offs (i + 1)) :: Int
+        ]
+  where
+    Dir {..} = rawPageDirectory page
+
+rawPageValueAt :: RawPage -> Int -> P.Vector Word8
+rawPageValueAt page@(RawPage off ba) i = do
+    P.Vector (mul2 off + start) (end - start) ba
+  where
+    offs  = rawPageValueOffsets page
+    start = fromIntegral (P.unsafeIndex offs i) :: Int
+    end   = fromIntegral (P.unsafeIndex offs (i + 1)) :: Int
+
+-- we could create unboxed array.
```

Comment:
Yeah, it would match the representation of a unboxed (not primitive) vector of pairs. That'd be quite neat and fit with the style you're using elsewhere. But just a neatness bonus.